### ISSUE TITLE
ClassGraph assumes that URLs inside URLClassLoader are all file://

### ISSUE THREAD

**Issue by chrisr3** [classgraph/classgraph#384]
Java's `URL` class is extensible via custom `URLStreamHandler` factories. However, ClassGraph's `URLClassLoaderHandler` assumes that all URLs are `file://` URLs.
```java
for (final URL url : urls) {
    if (url != null) {
        classpathOrder.addClasspathEntry(url.toString(), classLoader, scanSpec, log);
    }
}
```
The more general approach would involve reading the contents using `URL.openConnection()` or `URL.openStream()`, which would reasonably return something resembling either `ZipInputStream` or `JarInputStream`.

**Comment by lukehutch** [comment]
Did you actually test this? If an http/https URL doesn't work, it's a bug.

https://github.com/classgraph/classgraph/blob/master/src/main/java/nonapi/io/github/classgraph/fastzipfilereader/NestedJarHandler.java#L231

One of the unit tests that is always run involves fetching a jar over https.

https://github.com/classgraph/classgraph/blob/2c369f3996bd3ba3b18754face4c9f9d1f0503b0/src/test/java/io/github/classgraph/issues/issue128/Issue128Test.java#L67

The code you linked just converts any `URL` passed into ClassGraph into its string representation (which is another valid way to provide URLs on the classpath), they don't have to use an actual `URL` object).

**Comment by chrisr3** [comment]
I have defined a custom URL scheme (i.e. neither `file://`, `jar://`, `http://` nor `https://`) and confirmed that this does not work, yes.

**Comment by chrisr3** [comment]
Just in case this is relevant, here is how I am creating my `ClassGraph`:
```kotlin
val classGraph = ClassGraph()
    .overrideClassLoaders(appLoader)
    .ignoreParentClassLoaders()
    .disableDirScanning()
    .enableExternalClasses()
    .enableAllInfo()
    .verbose(true)
```

**Comment by lukehutch** [comment]
Hmm. I didn't even know it was possible to create custom URL schemes...

Code will have to be added to ClassGraph to support this, yes. I'll look into it later today.

**Comment by chrisr3** [comment]
FWIW, here are the classes I used:
```kotlin
package com.example.memory

import java.io.FileNotFoundException
import java.io.IOException
import java.io.InputStream
import java.net.*
import java.nio.ByteBuffer
import java.util.*
import java.util.zip.ZipInputStream
import kotlin.math.min

data class MemoryURL(val value: URL, val lock: String)

class MemoryURLStreamHandler(val scheme: String) : URLStreamHandler() {
    private val dataCache: MutableMap<String, ByteBuffer> = WeakHashMap()

    @Throws(MalformedURLException::class)
    fun createURL(path: String, data: ByteBuffer): MemoryURL {
        if (path.contains("!/")) {
            throw MalformedURLException("Invalid URL path '$path'")
        }

        val url = URL(scheme, "", -1, path, this)
        val lock = url.toString()
        if (dataCache.putIfAbsent(lock, data.asReadOnlySlice()) != null) {
            throw MalformedURLException("URL '$lock' already exists")
        }
        return MemoryURL(url, lock)
    }

    private fun ByteBuffer.asReadOnlySlice(): ByteBuffer {
        return (if (isReadOnly) this else asReadOnlyBuffer()).slice()
    }

    @Throws(IOException::class)
    override fun openConnection(url: URL): URLConnection {
        val urlValue = url.toString().split("!/")
        val data = dataCache[urlValue.first] ?: throw IOException("No data for URL '${urlValue.first}'")
        return Connection(url, urlValue.second, data.duplicate())
    }

    private fun String.split(delimiter: String): Pair<String, String> {
        val idx = indexOf(delimiter)
        return if (idx == -1) {
            Pair(this, "")
        } else {
            Pair(substring(0, idx), substring(idx + delimiter.length, length))
        }
    }

    private class Connection(url: URL, private val path: String, private val data: ByteBuffer) : URLConnection(url) {
        private val handler = ConnectionHandler()

        init {
            allowUserInteraction = false
            useCaches = false
        }

        @Throws(IOException::class)
        @Synchronized
        override fun connect() {
            if (!connected) {
                handler.connect()
                connected = true
            }
        }

        @Throws(IOException::class)
        override fun getInputStream(): InputStream {
            connect()
            return handler
        }

        @Throws(IOException::class)
        override fun getContentLengthLong(): Long {
            connect()
            return handler.dataSize
        }

        @Throws(IOException::class)
        override fun getContent(): ByteBuffer {
            connect()
            if (path.isNotEmpty()) {
                throw UnsupportedOperationException("Not supported for $url")
            }
            return data
        }

        @Throws(IOException::class)
        override fun getContentType(): String {
            connect()
            return "application/octet-stream"
        }

        override fun setDoOutput(doOutput: Boolean) = throw UnsupportedOperationException("Output not supported")
        override fun setAllowUserInteraction(allowUserInterfaction: Boolean) = throw UnsupportedOperationException("User interaction not supported")

        private inner class ConnectionHandler : InputStream() {
            @Volatile
            private var dataStream: InputStream = ByteBufferInputStream(data)

            @Volatile
            var dataSize: Long = data.limit().toLong()
                private set

            @Throws(FileNotFoundException::class)
            fun connect() {
                if (path.isNotEmpty()) {
                    val zipStream = ZipInputStream(dataStream)
                    while (true) {
                        val entry = zipStream.nextEntry ?: throw FileNotFoundException(path)
                        if (path == entry.name) {
                            dataStream = zipStream
                            dataSize = entry.size
                            break
                        }
                    }
                }
            }

            @Throws(IOException::class)
            override fun read(): Int = dataStream.read()

            @Throws(IOException::class)
            override fun read(buffer: ByteArray, offset: Int, size: Int): Int = dataStream.read(buffer, offset, size)

            @Throws(IOException::class)
            override fun skip(size: Long): Long = dataStream.skip(size)

            override fun available(): Int = dataStream.available()

            @Throws(IOException::class)
            override fun close() = dataStream.close()
        }
    }
}

private class ByteBufferInputStream(private val bytes: ByteBuffer) : InputStream() {
    private companion object {
        private const val EOF = -1
    }

    override fun read(): Int {
        return if (bytes.hasRemaining()) bytes.get().toInt() else EOF
    }

    override fun read(buffer: ByteArray, offset: Int, length: Int): Int {
        return if ((offset < 0) || (length < 0) || offset + length > buffer.size) {
            throw IndexOutOfBoundsException()
        } else {
            val bytesRemaining = bytes.remaining()
            if (bytesRemaining > 0) {
                val bytesRead = min(length, bytesRemaining)
                bytes.get(buffer, offset, bytesRead)
                bytesRead
            } else {
                EOF
            }
        }
    }

    override fun available(): Int = bytes.remaining()

    override fun markSupported(): Boolean = true

    override fun mark(readLimit: Int) {
        bytes.mark()
    }

    override fun reset() {
        bytes.reset()
    }
}
```

**Comment by lukehutch** [comment]
@chrisr3 can you please test the git master version? It should fix your issue.

**Comment by lukehutch** [comment]
BTW you will additionally need `.enableRemoteJarScanning()` before `.scan()` to enable scanning of any URL schemes other than `file:`.

**Comment by lukehutch** [comment]
Actually I'm pretty sure the fixes should work for you (there's a unit test that shows it works for me, at least), so I pushed out a new release with these changes, 4.8.56.

**Comment by lukehutch** [comment]
Assuming fixed in 4.8.56. Let me know if not, and I'll reopen. Thanks for the bug report!

**Comment by chrisr3** [comment]
> Assuming fixed in 4.8.56. Let me know if not, and I'll reopen. Thanks for the bug report!

Thanks, that was quick! I'll take it for a spin and let you know :smile:

**Comment by lukehutch** [comment]
Oops, I shouldn't have overconfidently assumed that my unit test passing was enough to verify that this worked... there were a few regressions. (It also broke scanning on Windows.) I'll try to get these fixed ASAP.

**Comment by lukehutch** [comment]
Fixed and released in 4.8.57.
